### PR TITLE
Integrate level-finance

### PR DESCRIPTION
**Pull request by guiltylotus**

### CODE REVIEW

File: pkg/source/level-finance/pools_list_updater.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	d.hasInitialized = true
+
+	return pools, nil, nil
+}
+
+func (d *PoolsListUpdater) init(ctx context.Context) ([]entity.Pool, error) {
+	var (
+		allAssets = make([]common.Address, 10)
```
Comment (ChauNguyen-K) [inline pkg/source/level-finance/pools_list_updater.go:49]: Why we hardcode 10 for all Assests, I see that ABI supports function method [getAllAssets](https://arbiscan.io/address/0x32B7bF19cb8b95C27E644183837813d4b595dcc6#readProxyContract)
`[ getAllAssets method Response ]
  assets   address[] :  
[[0x2f2a2543B76A4166549F7aaB2e75Bef0aefC5B0f]
[0x82aF49447D8a07e3bd95BD0d56f35241523fBab1]
[0x912CE59144191C1204E64559FE8253a0e49E6548]
[0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9]
[0xaf88d065e77c8cC2239327C5EDb3A432268e5831]]
  isStable   bool[] :  false,false,false,true,true`

Comment (ChauNguyen-K) [reply pkg/source/level-finance/pools_list_updater.go:49]: Your implementation is still ok, incase I didn't find getAllAssets method on BSC chain.

Comment (guiltylotus) [reply pkg/source/level-finance/pools_list_updater.go:49]: Yes, there are some different between bsc and arbitrum to get pool states. I updated in notion https://www.notion.so/kybernetwork/Integration-Notes-Level-Finance-Study-note-23dded16b3d444caacd84c253f8eaecf?pvs=4

File: pkg/source/level-finance/pool_tracker.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	calls.AddCall(&ethrpc.Call{
+		ABI:    LiquidityPoolAbi,
+		Target: p.Address,
+		Method: liquidityPoolMethodVirtualPoolValue,
+		Params: nil,
+	}, []interface{}{&virtualPoolValue})
+	calls.AddCall(&ethrpc.Call{
+		ABI:    LiquidityPoolAbi,
+		Target: p.Address,
+		Method: liquidityPoolMethodFee,
```
[the remark this answers is not in the document]
Comment (guiltylotus) [reply pkg/source/level-finance/pool_tracker.go]: I updated version for arbitrum

File: pkg/source/level-finance/math.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	if err := validateMaxLiquidity(tokenInInfo); err != nil {
+		return nil, err
+	}
+
+	return amountOutAfterFee, nil
+}
+
+func calcSwapOutput(tokenIn, tokenOut *TokenInfo, amountIn *big.Int, state *PoolState) (*big.Int, *big.Int, error) {
+	priceIn := new(big.Int).Set(tokenIn.MinPrice)
+	priceOut := new(big.Int).Set(tokenOut.MaxPrice)
```
Comment (ChauNguyen-K) [inline pkg/source/level-finance/math.go:44]: The same with priceIn above.

File: pkg/source/level-finance/math.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+	if err := validateMaxLiquidity(tokenInInfo); err != nil {
+		return nil, err
+	}
+
+	return amountOutAfterFee, nil
+}
+
+func calcSwapOutput(tokenIn, tokenOut *TokenInfo, amountIn *big.Int, state *PoolState) (*big.Int, *big.Int, error) {
+	priceIn := new(big.Int).Set(tokenIn.MinPrice)
```
[the remark this answers is not in the document]
Comment (guiltylotus) [reply pkg/source/level-finance/math.go:43]: In 1 block, the swap does not cause the price changes. So we can use directly getPrice result from the contract.

**Pull request closed without merging**